pilot/pkg/config/kube/crd/controller/controller.go

```diff
@@ -182,29 +182,37 @@ func (c *controller) createInformer(
 	handler := &kube.ChainHandler{}
 	handler.Append(c.notify)
 
+	// Wrap the list function with supplied validation function, so that invalid CRD does not enter store.
+	vlf := func(opts meta_v1.ListOptions) (result runtime.Object, err error) {
+		// List only returns error, if type is missing.
+		if result, err = lf(opts); err == nil {
+			// This check is primarily needed for tests which use mock config - but helps otherwise also.
+			if obj, ok := result.(crd.IstioObject); ok {
+				if err = vf(obj); err != nil {
+					key := obj.GetObjectMeta().Namespace + "/" + obj.GetObjectMeta().Name
+					log.Errorf("CRD validation failed: %s %s %v", obj.GetObjectKind().GroupVersionKind().GroupKind().Kind,
+						key, err)
+					k8sErrors.With(nameTag.Value(key)).Record(1)
+					return nil, err
+				}
+			}
+		}
+		return
+	}
+
 	// TODO: finer-grained index (perf)
 	informer := cache.NewSharedIndexInformer(
-		&cache.ListWatch{ListFunc: lf, WatchFunc: wf}, o,
+		&cache.ListWatch{ListFunc: vlf, WatchFunc: wf}, o,
 		resyncPeriod, cache.Indexers{})
 
 	informer.AddEventHandler(
 		cache.ResourceEventHandlerFuncs{
 			// TODO: filtering functions to skip over un-referenced resources (perf)
 			AddFunc: func(obj interface{}) {
-				if err := vf(obj); err != nil {
```

Comment:
@ramaraochavali Doesn't the listed objects run this AddFunc too?

Reply:
ref: `k8s.io/client-go/tools/cache/controller.go#L360~L372`

```
			for _, d := range obj.(Deltas) {
				switch d.Type {
				case Sync, Added, Updated:
					if old, exists, err := clientState.Get(d.Object); err == nil && exists {
						if err := clientState.Update(d.Object); err != nil {
							return err
						}
						h.OnUpdate(old, d.Object)
					} else {
						if err := clientState.Add(d.Object); err != nil {
							return err
						}
						h.OnAdd(d.Object)
					}
```

When list returned, this callback will be triggered with `Sync` type.

Reply:
Yeah, I was just updating the above comment. I think we need validation in add and update as well as per this https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L399. I have added it back. Is that what you are also suggesting?

Reply:
I mean previously this should already be validated since @yxue 's pr

Reply:
But the initial listing does not call the callbacks as per here https://github.com/kubernetes/client-go/blob/master/tools/cache/reflector.go#L187 and https://github.com/kubernetes/client-go/blob/8d0e6f1b7b780ef8c8f81800d504509cd1f68e9f/tools/cache/reflector.go#L349 - It just replalces in the store

Reply:
Yeah, i doubt under which case the listed objects donot go through `processLoop `